crates/astria-sequencer/src/transaction/signed_transaction.rs

```diff
@@ -0,0 +1,155 @@
+use anyhow::{
+    anyhow,
+    bail,
+    Context as _,
+    Result,
+};
+use astria_proto::sequencer::v1::SignedTransaction as ProtoSignedTransaction;
+use async_trait::async_trait;
+use penumbra_storage::{
+    StateRead,
+    StateWrite,
+};
+use prost::Message as _;
+use tracing::instrument;
+
+use crate::{
+    accounts::{
+        transaction::Transaction as AccountsTransaction,
+        types::Address,
+    },
+    crypto::{
+        PublicKey,
+        Signature,
+        Verifier,
+    },
+    hash,
+    transaction::{
+        ActionHandler,
+        TransactionHash,
+        UnsignedTransaction,
+    },
+};
+
+#[derive(Debug, Clone)]
+pub struct SignedTransaction {
+    pub signature: Signature,
+    pub public_key: PublicKey,
+    pub transaction: UnsignedTransaction,
+}
+
+impl SignedTransaction {
+    /// Verifies the transaction signature.
+    /// The transaction signature message is the hash of the transaction.
+    ///
+    /// # Errors
+    ///
+    /// - If the signature is invalid
+    pub fn verify_signature(&self) -> Result<()> {
+        self.public_key
+            .verify(&self.transaction.hash(), &self.signature)
+            .context("failed to verify transaction signature")
+    }
+
+    /// Returns the address which signed the transaction.
+    ///
+    /// # Errors
+    ///
+    /// - If the public key cannot be converted into an address
+    pub fn signer_address(&self) -> Result<Address> {
+        Address::try_from(&self.public_key)
+    }
+
+    #[must_use]
+    pub fn to_vec(&self) -> Vec<u8> {
+        let proto = ProtoSignedTransaction {
+            transaction: Some(match &self.transaction {
+                UnsignedTransaction::AccountsTransaction(tx) => {
+                    astria_proto::sequencer::v1::UnsignedTransaction {
+                        value: Some(astria_proto::sequencer::v1::unsigned_transaction::Value::AccountsTransaction (
```

Comment:
what's the difference between `unsigned_transaction` here and `UnsignedTransaction` above?

Reply:
when prost generates this type:
```
message UnsignedTransaction {
    oneof value {
        AccountsTransaction accounts_transaction = 1;
    }   
}
```

it creates a module `unsigned_transaction` as follows to deal with the `oneof`:
```
/// Nested message and enum types in `UnsignedTransaction`.
pub mod unsigned_transaction {
    #[allow(clippy::derive_partial_eq_without_eq)]
#[derive(Clone, PartialEq, ::prost::Oneof)]
    pub enum Value {
        #[prost(message, tag="1")]
        AccountsTransaction(super::AccountsTransaction),
    }
}
```

`UnsignedTransaction` is a struct that contains a `Value`:
```
pub struct UnsignedTransaction {
    #[prost(oneof="unsigned_transaction::Value", tags="1")]
    pub value: ::core::option::Option<unsigned_transaction::Value>,
}
```